rpc/dns_monitor.go

```diff
@@ -0,0 +1,203 @@
+
+	// Prevent concurrent reconnections - global state independent of monitor lifecycle
+	reconnectionInProgress atomic.Bool
+
+	// Rate limiting state - global to survive monitor restarts during reconnection
+	lastReconnectTime  time.Time
+	lastReconnectMutex sync.Mutex
+)
+
+// StartDNSMonitor initializes and starts the background DNS monitor
+func StartDNSMonitor(enabled bool, checkInterval int, connectionString string) {
+	dnsMonitorLock.Lock()
+	defer dnsMonitorLock.Unlock()
+
+	// Stop existing monitor if running
+	if dnsMonitor != nil {
+		stopDNSMonitorInternal()
+	}
+
+	// Don't start if disabled
+	if !enabled {
+		Log.Debug("DNS monitor is disabled")
+		return
+	}
+
+	// Validate configuration
+	if connectionString == "" {
+		Log.Warning("DNS monitor enabled but connection string is empty, skipping monitor start")
+		return
+	}
+
+	// Set default interval if not specified or invalid
+	if checkInterval <= 0 {
+		checkInterval = 30 // Default to 30 seconds
+	}
+
+	// Create context for lifecycle management
+	ctx, cancel := context.WithCancel(context.Background())
+
+	dnsMonitor = &DNSMonitor{
+		enabled:       true,
+		checkInterval: time.Duration(checkInterval) * time.Second,
+		connectionStr: connectionString,
+		ctx:           ctx,
+		cancel:        cancel,
+		stopComplete:  make(chan struct{}),
+	}
+
+	Log.WithFields(logrus.Fields{
+		"check_interval": checkInterval,
+		"connection":     connectionString,
+	}).Info("Starting background DNS monitor for MDCB connection")
+
+	// Start the monitor loop in background
+	go dnsMonitor.monitorLoop()
+}
+
+// StopDNSMonitor gracefully stops the DNS monitor
+func StopDNSMonitor() {
+	dnsMonitorLock.Lock()
+	defer dnsMonitorLock.Unlock()
+
+	stopDNSMonitorInternal()
+}
+
+// stopDNSMonitorInternal stops the DNS monitor without acquiring the lock
+// (assumes caller already holds the lock)
+func stopDNSMonitorInternal() {
+	if dnsMonitor == nil {
+		return
+	}
+
+	Log.Info("Stopping background DNS monitor")
+
+	// Signal shutdown
+	dnsMonitor.cancel()
+
+	// Wait for monitor loop to complete (with timeout)
+	select {
+	case <-dnsMonitor.stopComplete:
+		Log.Debug("DNS monitor stopped gracefully")
+	case <-time.After(5 * time.Second):
+		Log.Warning("DNS monitor stop timed out after 5 seconds")
+	}
+
+	dnsMonitor = nil
+}
+
+// monitorLoop is the main loop that periodically checks DNS
+func (m *DNSMonitor) monitorLoop() {
+	defer close(m.stopComplete)
+
+	ticker := time.NewTicker(m.checkInterval)
+	defer ticker.Stop()
+
+	Log.Debug("DNS monitor loop started")
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			Log.Debug("DNS monitor loop received shutdown signal")
+			return
+		case <-ticker.C:
+			m.checkDNS()
+		}
+	}
+}
+
+// checkDNS performs the DNS check and reconnects if needed
+func (m *DNSMonitor) checkDNS() {
+	// Don't check if we're in emergency mode
+	if values.GetEmergencyMode() {
+		Log.Debug("DNS monitor: skipping check (emergency mode active)")
+		return
+	}
```

Comment:
why dont check if we are in emergency? am wondering the situation in which we fell into emergency because dns has changed then we do not check?